lib/profileAgents.js

```diff
@@ -602,6 +609,104 @@ async function _updateMeterController({meterId, controller}) {
   }
 }
 
+async function _getProfileAgentRecords({accountId, includeSecrets}) {
+  const query = {'profileAgent.account': accountId};
+  const projection = {_id: 0};
+  const collection = getCollection('profile-profileAgent');
+  let records = await collection.find(query, {projection}).toArray();
+  const zcapRefreshed = await _refreshProfileAgentZcaps({
+    profileAgentRecords: records
+  });
+  if(zcapRefreshed) {
+    records = await collection.find(query, {projection}).toArray();
+  }
+  if(!includeSecrets) {
+    records = await removeSecretsFromRecords({records});
+  }
+  return records;
+}
+
+async function _refreshProfileAgentZcaps({profileAgentRecords}) {
+  let zcapRefreshed = false;
+  const records = klona(profileAgentRecords);
+  const now = Date.now();
+  for(const record of records) {
+    if(!record.profileAgent.zcaps) {
+      continue;
+    }
+    const {zcaps} = record.profileAgent;
+    for(const zcapName in zcaps) {
+      if(zcapName !== 'profileCapabilityInvocationKey') {
+        const zcap = zcaps[zcapName];
+        const {expires} = zcap;
+        // check if the zcap is close to expiry
+        const expTime = Date.parse(expires);
+        const {zcapRefreshThreshold} = config.profile.profileAgent;
+        if(expTime - now <= zcapRefreshThreshold) {
+          if(zcapName === 'userDocument') {
+            await _refreshProfileAgentUserDocZcaps({zcap, now, record});
+          }
+          zcap.expires = new Date(now + DEFAULT_PROFILE_AGENT_ZCAP_TTL);
+          zcapRefreshed = true;
+        }
+      }
+    }
+    if(zcapRefreshed) {
+      // update the profile agent record
+      record.profileAgent.sequence += 1;
+      await update({profileAgent: record.profileAgent});
+    }
+  }
+  return zcapRefreshed;
+}
+
+async function _refreshProfileAgentUserDocZcaps({zcap, now, record} = {}) {
+  let zcapRefreshed = false;
+  const {zcapRefreshThreshold} = config.profile.profileAgent;
+  const kmsClient = new KmsClient({httpsAgent});
+  const profileSigner = await getProfileSigner({
+    kmsClient, profileAgentRecord: record
+  });
+  const docUrl = new URL(zcap.invocationTarget);
+  const edvId = `${docUrl.protocol}//${docUrl.hostname}:${docUrl.port}` +
+    `${docUrl.pathname.split('/').slice(0, 3).join('/')}`;
```

Comment:
Can this be determined some other way? This looks complex. I think we can just get the prefix that occurs before `/documents` or something like that -- there should be other code that does that or that uses a helper somewhere to mirror.

Reply:
Using a simpler helper function  to parse `edvId` from the zcap [3ddcb99](https://github.com/digitalbazaar/bedrock-profile/pull/99/commits/3ddcb994f53545e39125ae7f39a7bad0f0fed261)